### PR TITLE
CloudType UI Views

### LINKED ISSUE #5721: CloudType model
Create the CloudType model which is used similarly to DeviceType within the cloud app to designate type definitions for several other models such as Cloud Services.

Cloud Type:
- name
- description
- provider (FK to manufacturer)
- Extras config schema (json field)
- content types

### PR DESCRIPTION
**Pull request by timizuoebideri1**
# Closes #5721 
# What's Changed

# Screenshots
Screenshot 2024-06-11 at 7 15 16 AM
Screenshot 2024-06-11 at 7 15 31 AM

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: changes/5721.added [hunk 1]
```diff
@@ -1 +1 @@
-Added CloudType Model.
\ No newline at end of file
+Added CloudType Model and UI.
```
Comment (HanlinMiao) [inline changes/5721.added:1]: This should be similar to `5719.added`

File: nautobot/cloud/forms.py [hunk 2]
```diff
@@ -69,3 +71,69 @@ class CloudAccountFilterForm(NautobotFilterForm):
         queryset=Manufacturer.objects.all(), to_field_name="name", required=False
     )
     tags = TagFilterField(model)
+
+
+#
+# Cloud Account
+#
```
Comment (HanlinMiao) [inline nautobot/cloud/forms.py:78]:
```suggestion
#
# Cloud Type
#
```

File: nautobot/cloud/models.py [hunk 3]
```diff
@@ -7,7 +7,11 @@
 from nautobot.core.constants import CHARFIELD_MAX_LENGTH
 from nautobot.core.models import BaseModel
 from nautobot.core.models.generics import PrimaryModel
-from nautobot.extras.utils import extras_features
+from nautobot.extras.utils import CloudTypeModelsQuery, extras_features
+
+
+class CloudTypeCompatibleMixin:
```
Comment (HanlinMiao) [inline nautobot/cloud/models.py:13]: ~~How does this mixin work in practice? I am confused since there is no class attribute or anything.~~

Comment (HanlinMiao) [inline nautobot/cloud/models.py:13]: Also I **think** `CloudTypeMixin` should be sufficient.

File: nautobot/cloud/navigation.py [hunk 4]
```diff
@@ -30,6 +30,22 @@
                             ),
                         ),
                     ),
+                    NavMenuItem(
+                        link="cloud:cloudtype_list",
+                        name="Cloud Types",
+                        weight=100,
```
Comment (HanlinMiao) [inline nautobot/cloud/navigation.py:36]:
```suggestion
                        weight=200,
```

File: nautobot/cloud/tests/test_views.py [hunk 5]
```diff
@@ -33,3 +35,32 @@ def setUpTestData(cls):
...
+
+class CloudTypeTestCase(ViewTestCases.PrimaryObjectViewTestCase):
+    model = CloudType
+
+    @classmethod
+    def setUpTestData(cls):
+        providers = Manufacturer.objects.all()
+        CloudType.objects.create(name="Deletable Cloud Account 1", account_number="123456", provider=providers[0])
+        CloudType.objects.create(name="Deletable Cloud Account 2", account_number="234567", provider=providers[0])
+        CloudType.objects.create(name="Deletable Cloud Account 3", account_number="345678", provider=providers[0])
```
Comment (HanlinMiao) [inline nautobot/cloud/tests/test_views.py:48]: `CloudType` shouldn't have `account_number` as an attribute

File: nautobot/extras/utils.py [hunk 6]
```diff
@@ -219,6 +219,26 @@ def list_subclasses(self):
...
+        Return a list of classes that inherits from CloudTypeCompatibleMixin
+        """
+        # Avoid circular imports
+        from nautobot.cloud.models import CloudTypeCompatibleMixin
+
+        model_classes = []
+        for model_class in apps.get_models():
+            if issubclass(model_class, CloudTypeCompatibleMixin):
+                model_classes.append(model_class)
+        return model_classes
```
Comment (HanlinMiao) [inline nautobot/extras/utils.py:239]: Ahhhh I see how it works now

File: nautobot/cloud/models.py [hunk 7]
```diff
@@ -18,7 +22,7 @@
     "graphql",
     "webhooks",
 )
-class CloudAccount(PrimaryModel):
+class CloudAccount(CloudTypeCompatibleMixin, PrimaryModel):
```
Comment (glennmatthews) [inline nautobot/cloud/models.py:25]: I think CloudType `content_types` is meant to be for CloudNetwork and CloudService only at present? CloudAccount doesn't have `extra_config` so the schema defined by a CloudType isn't really relevant.

File: nautobot/extras/utils.py (hunk 6 above)
Comment (glennmatthews) [reply nautobot/extras/utils.py:239]: I'm not a big fan of us adding new `FooModelsQuery` classes. Can we not instead follow the pattern used by contacts, saved views, and static groups to be supportable by a general `FeatureQuery("cloud_types")` or similar?

File: nautobot/cloud/templates/cloud/cloudtype_retrieve.html [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+            <tr>
+                <td>Provider</td>
+                <td>{{ object.provider|hyperlinked_object }}</td>
+            </tr>
+            <tr>
+                <td>Content Type(s)</td>
+                <td>
+                    {% if object.content_types.all %}
+                        <ul>
+                            {% for content_type in object.content_types.all|dictsort:"model"|dictsort:"app_label" %}
```
Comment (glennmatthews) [inline nautobot/cloud/templates/cloud/cloudtype_retrieve.html:29]: I'm surprised the `dictsort` are needed, aren't content-types normally ordered by app-label and model by default?

File: nautobot/cloud/models.py (hunk 7 above)
Comment (timizuoebideri1) [reply nautobot/cloud/models.py:25]: There is no model `CloudService` at present

Comment (glennmatthews) [reply nautobot/cloud/models.py]: I'm aware, yes :). So in the immediate term it would be only `CloudNetwork`.

**Review by glennmatthews — APPROVED** [review]

**Pull request merged**